lib/system/nuttx/irq.c

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright (c) 2018, Pinecone Inc. and Contributors. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+/*
+ * @file	nuttx/irq.c
+ * @brief	NuttX libmetal irq definitions.
+ */
+
+#include <errno.h>
+#include <metal/irq.h>
+#include <metal/alloc.h>
+#include <nuttx/irq.h>
+
+/** IRQ handler descriptor structure */
+struct metal_irq_hddesc {
+	metal_irq_handler hd;     /**< irq handler */
+	void *drv_id;             /**< id to identify the driver
+	                               of the irq handler */
+};
+
+static int metal_irq_isr(int irq, void *context, void *arg);
+
+int metal_irq_register(int irq,
+		       metal_irq_handler hd,
+		       struct metal_device *dev,
+		       void *drv_id)
+{
+	struct metal_irq_hddesc *desc;
+	int ret;
+
+	if ((drv_id == NULL) || (hd == NULL))
+		return -EINVAL;
+
+	desc = metal_allocate_memory(sizeof(struct metal_irq_hddesc));
+	if (desc == NULL)
+		return -ENOMEM;
+
+	desc->hd = hd;
+	desc->drv_id = drv_id;
+
+	ret = irq_attach(irq, metal_irq_isr, desc);
```

Comment:
Here is the free process:
1.metal_irq_unregister trigger a fake interrupt by irq_dispatch
2.metal_irq_isr found that it's a fake interrupt and call irqchain_detach and sched_kfree
It's a little bit hack, since NuttX:
1.Already support the irq chain concept, I want to utilize it to save the coding
2.Three argument(irq, callback and arg) for irqchain_detach need exactly match irq_attach, so irq subsystem could just remove the right one from the list.
3.Since in metal_irq_register, I dynamiclly allocate metal_irq_hddesc, the normally approach could save it in a list and then we can find it in metal_irq_unregister.
4.But this approach make the implementation complicated and waste memory, so I choice a more clever(or hack) approach:
1.Trigger a fake interrupt by software
2.Irq subsystem find the right metal_irq_hddesc
3.Pass the descriptor to metal_irq_isr
4.metal_irq_isr call irqchain_detach with all right information
5.Free the descriptor by sched_kfree

Reply:
Sure, the irq dispatcher don't check the enable bit when dispatch the incoming interrupt(either real or fake).